stream-chat-android-client/src/main/java/io/getstream/chat/android/client/socket/experimental/ChatSocket.kt

```diff
@@ -0,0 +1,292 @@
+                }
+                false -> chatSocketStateService.onNetworkNotAvailable()
+            }
+        }
+
+        fun reconnect(connectionConf: SocketFactory.ConnectionConf) {
+            connectUser(connectionConf.asReconnectionConf())
+        }
+
+        return coroutineScope.launch {
+            chatSocketStateService.observer { state ->
+                when (state) {
+                    is State.RestartConnection -> {
+                        connectionConf?.let { chatSocketStateService.onReconnect(it) }
+                    }
+                    is State.Connected -> {
+                        healthMonitor.ack()
+                        callListeners { listener -> listener.onConnected(state.event) }
+                    }
+                    is State.Connecting -> {
+                        callListeners { listener -> listener.onConnecting() }
+                        when (state.isReconnection) {
+                            true -> reconnect(state.connectionConf.asReconnectionConf())
+                            false -> connectUser(state.connectionConf)
+                        }
+                    }
+                    is State.Disconnected -> {
+                        val disconnectCause: DisconnectCause = when (state) {
+                            is State.Disconnected.DisconnectedByRequest -> {
+                                healthMonitor.stop()
+                                coroutineScope.launch { disposeObservers() }
+                                streamWebSocket?.close()
+                                DisconnectCause.ConnectionReleased
+                            }
+                            is State.Disconnected.NetworkDisconnected -> {
+                                healthMonitor.stop()
+                                DisconnectCause.NetworkNotAvailable
+                            }
+                            is State.Disconnected.Stopped -> {
+                                healthMonitor.stop()
+                                disposeNetworkStateObserver()
+                                DisconnectCause.ConnectionReleased
+                            }
+                            is State.Disconnected.DisconnectedPermanently -> {
+                                healthMonitor.stop()
+                                coroutineScope.launch { disposeObservers() }
+                                DisconnectCause.UnrecoverableError(state.error)
+                            }
+                            is State.Disconnected.DisconnectedTemporarily -> {
+                                healthMonitor.onDisconnected()
+                                DisconnectCause.Error(state.error)
+                            }
+                            is State.Disconnected.WebSocketEventLost -> {
+                                connectionConf?.let { chatSocketStateService.onReconnect(it) }
+                                DisconnectCause.WebSocketNotAvailable
+                            }
```

Comment:
What do you think about extracting 2 functions from this `when` statement just for the sake of readability?
```kotlin
                    is State.Disconnected -> {
                        handleDisconnectedState(state)
                        val disconnectCause = state.toDisconnectCause()
                        callListeners { listener -> listener.onDisconnected(cause = disconnectCause) }
                    }
```

Mapper:
```kotlin
    private fun State.Disconnected.toDisconnectCause() = when (this) {
        is State.Disconnected.DisconnectedByRequest -> DisconnectCause.ConnectionReleased
        is State.Disconnected.NetworkDisconnected -> DisconnectCause.NetworkNotAvailable
        is State.Disconnected.Stopped -> DisconnectCause.ConnectionReleased
        is State.Disconnected.DisconnectedPermanently -> DisconnectCause.UnrecoverableError(error)
        is State.Disconnected.DisconnectedTemporarily -> DisconnectCause.Error(error)
        is State.Disconnected.WebSocketEventLost -> DisconnectCause.WebSocketNotAvailable
    }
```

Handling
```kotlin
    private fun handleDisconnectedState(state: State.Disconnected) {
        when (state) {
            is State.Disconnected.DisconnectedByRequest -> {
                healthMonitor.stop()
                coroutineScope.launch { disposeObservers() }
                streamWebSocket?.close()
            }
            is State.Disconnected.NetworkDisconnected -> {
                healthMonitor.stop()
            }
            is State.Disconnected.Stopped -> {
                healthMonitor.stop()
                disposeNetworkStateObserver()
            }
            is State.Disconnected.DisconnectedPermanently -> {
                healthMonitor.stop()
                coroutineScope.launch { disposeObservers() }
            }
            is State.Disconnected.DisconnectedTemporarily -> {
                healthMonitor.onDisconnected()
                DisconnectCause.Error(state.error)
            }
            is State.Disconnected.WebSocketEventLost -> {
                connectionConf?.let { chatSocketStateService.onReconnect(it) }
            }
        }
    }
```

Reply:
Make sense.
Let me update it